modules/apigee_m10n_teams/tests/src/Kernel/Entity/PurchasedRatePlanEntityKernelTest.php

```diff
@@ -0,0 +1,256 @@
+
+    $request = Request::create($url->toString(), 'GET');
+    $response = $this->container->get('http_kernel')->handle($request);
+
+    static::assertSame(Response::HTTP_FORBIDDEN, $response->getStatusCode());
+  }
+
+  /**
+   * Test team member can view purchased plans, and purchase.
+   *
+   * @throws \Exception
+   */
+  public function testPurchasedRatePlanEntityHasAccess() {
+    $this->createCurrentUserSession($this->member);
+
+    $new_permissions = [
+      'view purchased_plan' => 1,
+      'update purchased_plan' => 1,
+      'purchase rate_plan' => 1,
+      'view rate_plan' => 1,
+    ] + $this->getDefaultTeamRolePermissions();
+    $this->team_role_storage->changePermissions(TeamRoleInterface::TEAM_MEMBER_ROLE, $new_permissions);
+
+    $this->memberViewPurchasedRatePlans();
+    $this->memberCancelRatePlanPage();
+    $this->memberPurchaseRatePlanPage();
+  }
+
+  /**
+   * Test member with permission to view team purchased rate plans.
+   */
+  protected function memberViewPurchasedRatePlans() {
+    $url = Url::fromRoute('entity.purchased_plan.team_collection', [
+      'team' => $this->team->id(),
+    ]);
+    static::assertSame("/teams/{$this->team->id()}/monetization/purchased-plans", $url->toString());
+
+    $this->stack
+      ->queueMockResponse(['get_company_purchased_plans' => ['purchased_plans' => [$this->purchased_plan->decorated()]]]);
+
+    $request = Request::create($url->toString(), 'GET');
+    $response = $this->container->get('http_kernel')->handle($request);
+
+    $this->setRawContent($response->getContent());
+    static::assertSame(Response::HTTP_OK, $response->getStatusCode());
+
+    // Checking "Active and Future Purchased Plans" table columns.
+    $this->assertCssElementText('.purchased-plan-row:nth-child(1) td.purchased-plan-status span', 'Active');
+    $this->assertCssElementText('.purchased-plan-row:nth-child(1) td.purchased-plan-rate-plan a', $this->rate_plan->getDisplayName());
+    $this->assertCssElementText('.purchased-plan-row:nth-child(1) td.purchased-plan-start-date div', $this->purchased_plan->getStartDate()->format('m/d/Y'));
+    $this->assertCssElementText('.purchased-plan-row:nth-child(1) td.purchased-plan-end-date', '');
+  }
+
+  /**
+   * Purchase rate plan page with member with permission.
+   */
+  protected function memberPurchaseRatePlanPage() {
+    $url = Url::fromRoute('entity.rate_plan.team_purchase', [
+      'team' => $this->team->id(),
+      'product_bundle' => $this->product_bundle->id(),
+      'rate_plan' => $this->rate_plan->id(),
+    ]);
+    static::assertSame("/teams/{$this->team->id()}/monetization/product-bundle/{$this->product_bundle->id()}/plan/{$this->rate_plan->id()}/purchase", $url->toString());
+
+    $request = Request::create($url->toString(), 'GET');
+    $response = $this->container->get('http_kernel')->handle($request);
+
+    $this->setRawContent($response->getContent());
```

Comment:
Is this line needed? I do not see how it is used in rest of method.

Reply:
Removed :)